### ISSUE TITLE
Disabling Basset Completely

### ISSUE THREAD

**Issue by bretvanhorn** [Laravel-Backpack/basset#50]
Hi-

TL;DR: Is there a way to completely disable basset in Backpack 6.x?

I have been trying to upgrade Backpack 5.x to 6.x and have had repeated issues with permissions errors when we load a page for the first time and basset tries to create a new directory in the storage path. The issue seems to be that my webserver is running under the www-data user, but the unix user for the account under which the site is running under Laravel Forge is 'forge.' I have added some commands to our deploy script to set default user/groups/permissions on the directory, but that feels a bit hacky to me. 

And yes, I am running basset:clear and basset:internalize. I have tried running basset:cache referenced in the docs, but that command does not exist. 

To make it more challenging, I am seeing some assets not being cached at all, either by running "basset:internalize" or loading the page. Examples are things like Backpack CRUD buttons, so the delete button js (//storage/backpack/crud/buttons/delete-button-en-7ff74202.js?0c1862722e18, for example) loads as a 404, and then trying to delete an entry from a CRUD list view results in a "deleteEntry not found" error in the console. Enabling dev mode fixes it on my local, yet breaks the site (no assets load) on our staging server. 

Honestly, at this point, I would love to just disable basset completely so we can move on with this deployment and come back to integrating it later, and I know I can revert the @basset tags to <script> and <link> tags for my own assets, but will it still continue to try loading Backpack assets from the storage link?

Between these issues and many of our models' admin panels not working completely with the Backpack 6.x upgrade, anything we can do to simplify things is most welcome. 

Composer Versions (ran composer update earlier this morning, as well):

"php": "8.2.*",
"laravel/framework": "^10.0",
"backpack/crud": "^6.0.0",
"backpack/pro": "^6.0.0",
"backpack/permissionmanager": "^6.0.16",
"backpack/theme-tabler": "dev-main",
"backpack/theme-coreuiv2": "dev-main"`
"backpack/generators": "v6.x-dev",
 "backpack/basset": "^0.16.0",

**Comment by pxpm** [comment]
Hey @bretvanhorn thanks for reaching 🙏 .

Being in beta is important in this regard. I don't expect the process of migration to be done "all-in" in this phase, but more people migrating and telling what issues they find make us come up with solutions to make the "final push" when v6 is released smoother. 

Basset was a big big change. I am hoping we didn't introduced any major changes. I haven't got the problem you reported yet, but my time may be around the corner, so I would like to cover the most possible scenarios before the launch. 

@promatik was the one working on Bassets, I am not too deep in the internals, you can enable dev mode as you said to don't internalize assets. Isn't that working for you ? 
Want to report some specific case where it's broken? Maybe we can fix it, less one headache for you 👍 

 `theoretically` you wouldn't need to change nothing in your blade files, basset should be capable of using the url/content you provided and serve the content without "internalizing". 

Basset will only internalize what's inside `@basset` and `@bassetBlock`. 

> Between these issues and many of our models' admin panels not working completely with the Backpack 6.x upgrade, anything we can do to simplify things is most welcome.

Want to share some of those errors ? Are they related with the basset issue ? Can we make something else to ease out the migration ? 

I don't mind creating a migration script whatever that's necessary. Making the upgrade process easier seems to be on both of our minds. 

Please post some error traces, some pictures, some something. I want to help you out, but I need to know what direction to take it 👍 

Cheers

**Comment by bretvanhorn** [comment]
Hey @pxpm! Thank you for the thoughtful reply. 

I was able to get things to a place where everything is working by removing our custom scripts and css files from basset, and putting the site into dev mode. That gets us where we need to be for now. Speed is not a huge issue as it is a private web app that sits behind a VPN and has a small user base. 

As for the specific errors, it was a general "unable to create file" error in Laravel, but the root cause was file and directory permissions between the forge user (under which PHP CLI runs) and the www-data user, under which nginx runs. My temp fix was to set the default owner under the storage/app directory to forge, the default group to www-data, and set the sticky bit permissions to 775. I don't have to do that anymore with the way things are running. so I feel better about that as well. 

Strangely this permissions error happened sporadically on some files, and if basset processed a CSS file that has relative font paths, it would fail to copy the fonts over, causing 404 errors on some webfonts. 

I do admit freely we are running a lot of legacy code, so it presents perhaps a messier upgrade process for us. I do not mean to put all of the blame on basset. On our list of longer-term goals is to rebuild the UX for this app, at which time we would probably try to start with a fresh Backpack install. 

Backpack is great, it has saved us countless hours, so I am happy to keep upgrading. I just have some time constraints with this sprint and need to get it to where it works.

**Comment by tabacitu** [comment]
Thanks for letting us know @bretvanhorn !

> Strangely this permissions error happened sporadically on some files, and if basset processed a CSS file that has relative font paths, it would fail to copy the fonts over, causing 404 errors on some webfonts.

@promatik do you see any scenario where Basset could internalize assets using a different permission/user/etc? What do you could be the cause of the above?

**Comment by tabacitu** [comment]
🎉🙏 Can I use that as a testimonial on our website? 👀

**Comment by promatik** [comment]
> _@promatik do you see any scenario where Basset could internalize assets using a different permission/user/etc? What do you could be the cause of the above?_

I can't see how this happens, all the "disk->put" commands are done on the same file, always by a running instance of php, the user must be the same, always.

---

> _The issue seems to be that my webserver is running under the www-data user, but the unix user for the account under which the site is running under Laravel Forge is 'forge.'_

I'll run a test with Forge right now 👌
I'll try to see if I have basset files under the `www-data` user.

**Comment by tabacitu** [comment]
Same here - I tried to see on multiple DO droplets managed with Forge and...  

<img width="1031" alt="CleanShot 2023-06-24 at 12 34 20@2x" src="https://github.com/Laravel-Backpack/basset/assets/1032474/a8c97d2a-9666-41df-a6c3-1eafa94d5957">

Looks like something's off in your server configuration then? I'm skeptical to say it's a scenario the package should cater to.

---

That being said, I do think it's a good idea to have an option to completely disable Basset. But AFAIK that's exactly what BASSET_DEV_MODE does. Right @promatik ? When that's enabled, what does Basset keep doing and what does it stop doing? I'm asking because... this isn't really documented 👀 https://github.com/Laravel-Backpack/basset/issues/38

**Comment by promatik** [comment]
True @tabacitu 👌

Basset cannot be **completely** disabled, because in some situations, the assets aren't available publicly at all, like online zip archives, or vendor files, these all must be **always** internalized.

Dev mode indeed disables basset for all the **possible** situations 👌

So, in my opinion, having a `BASSET_DISABLE` config variable that could be an alias to the `BASSET_DEV_MODE` is not a good idea, because it wouldn't be true and it might mislead devs into thinking it was fully disabled. 

In my opinion, if a developer wants to avoid internalizing files in the server, the only way is to go back to the previous way to deal with files, internalize in a public folder and commit them 😕

---

@tabacitu should we document it? Or should this thread be enough for future developers with the same problem to figure out how to solve it?

**Comment by tabacitu** [comment]
> So, in my opinion, having a BASSET_DISABLE config variable that could be an alias to the BASSET_DEV_MODE is not a good idea, because it wouldn't be true and it might mislead devs into thinking it was fully disabled.

Good thinking! I agree, that would probably add confusion.

> @tabacitu should we document it? Or should this thread be enough for future developers with the same problem to figure out how to solve it?

I think so too. This thread should be enough for people to understand WHY there is no way to completely disable Basset - thanks for the explanation. 

----

Let's close this issue, since I think @bretvanhorn 's problem here was a server misconfiguration that is unlikely to happen for a lot of people, and we DO have a way to disable as much as possible of Basset (`BASSET_DEV_MODE`). 

Is that reasonable @bretvanhorn ? Let us know if not and we'll reopen.

Cheers!
